pkg/specgen/generate/pod_create.go

```diff
@@ -2,22 +2,152 @@ package generate
+	"fmt"
+	"io/ioutil"
 	"net"
+	"os"
 
+	buildahDefine "github.com/containers/buildah/define"
 	"github.com/containers/common/pkg/config"
 	"github.com/containers/podman/v3/libpod"
 	"github.com/containers/podman/v3/libpod/define"
 	"github.com/containers/podman/v3/pkg/domain/entities"
 	"github.com/containers/podman/v3/pkg/rootless"
 	"github.com/containers/podman/v3/pkg/specgen"
+	"github.com/containers/podman/v3/pkg/specgenutil"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
+// builds an image with the local pause binary
+func buildInfraImage(rt *libpod.Runtime, rtConfig *config.Config) (string, error) {
+	// Make sure the name of the image includes the current version and the
+	// build time, so it gets rebuilt on updates.
+	version, err := define.GetVersion()
+	if err != nil {
+		return "", fmt.Errorf("getting version: %w", err)
+	}
+	pauseImageName := fmt.Sprintf("localhost/podman-pause:%s-%d", version.Version, version.Built)
+
+	// First, check whether we already built the image.
+	imageExists, err := rt.LibimageRuntime().Exists(pauseImageName)
+	if err != nil {
+		return "", err
+	}
+	if imageExists {
+		return pauseImageName, nil
+	}
+
+	pausePath, err := rtConfig.FindHelperBinary("pause", false)
+	if err != nil {
+		return "", fmt.Errorf("pause binary for infra container not found: %w", err)
+	}
+
+	buildInstructions := fmt.Sprintf(`FROM scratch
+ENTRYPOINT ["/pause"]
+COPY %s /pause`, pausePath)
+
+	tmpf, err := ioutil.TempFile("", "podman-pause.containerfile")
+	if err != nil {
+		return "", err
+	}
+	defer os.Remove(tmpf.Name())
+	defer tmpf.Close()
+
+	if _, err := tmpf.WriteString(buildInstructions); err != nil {
+		return "", err
+	}
+
+	buildOptions := buildahDefine.BuildOptions{
+		CommonBuildOpts: &buildahDefine.CommonBuildOptions{},
+		IIDFile:         "/dev/null", // To prevent Buildah from writing to stdout!
+		Output:          pauseImageName,
+		Quiet:           true,
+	}
+	if _, _, err := rt.Build(context.Background(), buildOptions, tmpf.Name()); err != nil {
+		return "", fmt.Errorf("building pause image for infra container: %w", err)
+	}
+
+	return pauseImageName, nil
+}
+
+func makeInfraSpec(p *entities.PodSpec, rt *libpod.Runtime) error {
+	if p.PodSpecGen.NoInfra {
+		return nil
+	}
+
+	rtConfig, err := rt.GetConfigNoCopy()
+	if err != nil {
+		return err
+	}
+
+	// Unless set to a non-default infra image (either by the user or
+	// containers.conf), build an infra image with the local pause binary.
+	imageName := p.PodSpecGen.InfraImage
+	if imageName == "" {
+		imageName = rtConfig.Engine.InfraImage
+	}
+	if imageName == "" || imageName == config.DefaultInfraImage {
+		name, err := buildInfraImage(rt, rtConfig)
+		if err != nil {
+			return err
+		}
+		imageName = name
+	} else {
+		_, err := rt.LibimageRuntime().Pull(context.Background(), imageName, config.PullPolicyMissing, nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	// The infra spec is *not* marshaled over the REST API, so we have to
```

Comment:
see, I think there is a discussion to be had about whether or not we should marshal over the api... adds some weird cases we would have to account for but also removes the necessity for this.